### PR TITLE
[Keyboard] Add avalanche

### PR DESCRIPTION
**Pull request by winder**
## Description

Add support for avalanche keyboard. This PR supersedes #14366 

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: keyboards/avalanche/v1/config.h [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+#pragma once
+
+#define VENDOR_ID       0xCEE2
+#define PRODUCT_ID      0x0001
+#define DEVICE_VER      0x0001
+#define MANUFACTURER    vitvlkv
+#define PRODUCT         Avalanche
+#define DESCRIPTION     QMK keyboard firmware for Avalanche
```
Comment (drashna) [inline keyboards/avalanche/v1/config.h:11]:
```suggestion
```

File: keyboards/avalanche/v1/v1.h [hunk 2]
```diff
@@ -0,0 +1,31 @@
+// Copyright 2022 vitvlkv (@vlkv)
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+#pragma once
+
+#include "../avalanche.h"
```
Comment (drashna) [inline keyboards/avalanche/v1/v1.h:6]:
```suggestion
#include "avalanche.h"
```

File: keyboards/avalanche/v2/config.h [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+#pragma once
+
+#define VENDOR_ID       0xCEE2
+#define PRODUCT_ID      0x0002
+#define DEVICE_VER      0x0002
+#define MANUFACTURER    vitvlkv
+#define PRODUCT         Avalanche
+#define DESCRIPTION     QMK keyboard firmware for Avalanche
```
Comment (drashna) [inline keyboards/avalanche/v2/config.h:11]:
```suggestion
```

File: keyboards/avalanche/v2/v2.h [hunk 4]
```diff
@@ -0,0 +1,31 @@
+// Copyright 2022 vitvlkv (@vlkv)
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+#pragma once
+
+#include "../avalanche.h"
```
Comment (drashna) [inline keyboards/avalanche/v2/v2.h:6]:
```suggestion
#include "avalanche.h"
```

File: keyboards/avalanche/v3/config.h [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+#pragma once
+
+#define VENDOR_ID       0xCEE2
+#define PRODUCT_ID      0x0003
+#define DEVICE_VER      0x0003
+#define MANUFACTURER    vitvlkv
+#define PRODUCT         Avalanche
+#define DESCRIPTION     QMK keyboard firmware for Avalanche
```
Comment (drashna) [inline keyboards/avalanche/v3/config.h:11]:
```suggestion
```

File: keyboards/avalanche/v3/v3.h [hunk 6]
```diff
@@ -0,0 +1,31 @@
+// Copyright 2022 vitvlkv (@vlkv)
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+#pragma once
+
+#include "../avalanche.h"
```
Comment (drashna) [inline keyboards/avalanche/v3/v3.h:6]:
```suggestion
#include "avalanche.h"
```

**Review by drashna — CHANGES_REQUESTED** [review]

**Review by noroadsleft — COMMENTED** [review]
Two general thoughts:

1. I'd much prefer if the license headers (and perhaps the USB `MANUFACTURER` strings as well) used the full name `Vitaly Volkov` instead of the abbreviated `vitvlkv`.
2. Unless there's some specific reason for the keymaps being in `avalanche/keymaps/default_v(1|2|3)`, a much cleaner implementation in my opinion would be to put each keymap at revision level:
   ```
   cd qmk_firmware/keyboards/avalanche/
   # create the destination directories:
   mkdir -p v1/keymaps/ v2/keymaps/ v3/keymaps/
   # move the keymap files:
   git mv keymaps/default_v1/ v1/keymaps/default/
   git mv keymaps/default_v2/ v2/keymaps/default/
   git mv keymaps/default_v3/ v3/keymaps/default/
   git commit
   # remove the now-empty `keyboards/avalanche/keymaps/` directory:
   git clean -f -- .
   ```
   Doing this means one specifies only the Avalanche revision (`avalanche/v1`, `avalanche/v2`, or `avalanche/v3`), and the keymap value is always `default`.

**Comment by winder** [comment]
Thanks for the feedback, PR is updated.

File: keyboards/avalanche/avalanche.c [hunk 7]
```diff
@@ -1,4 +1,4 @@
-// Copyright 2022 vitvlkv (@vlkv)
+// Copyright 2022 Viatly Volkov (@vlkv)
```
Comment (noroadsleft) [inline keyboards/avalanche/avalanche.c:1]: Typo.

```suggestion
// Copyright 2022 Vitaly Volkov (@vlkv)
```

File: keyboards/avalanche/readme.md [hunk 8]
```diff
@@ -4,17 +4,17 @@
...
 
-* Keyboard Maintainer: [vitvlkv](https://github.com/vlkv)
+* Keyboard Maintainer: [Vitaly Volkov](https://github.com/vlkv)
 * Hardware Supported: Avalanche PCB v1, v2, v3, Pro Micro 5V/16MHz and compatible.
 * Hardware Availability: [Avalanche](https://github.com/vlkv/avalanche).
 
 Make example for this keyboard (after setting up your build environment):
 
-    make avalanche:default_v3
+    make avalanche/v3:default
```
Comment (noroadsleft) [inline keyboards/avalanche/readme.md:13]:
```suggestion
    make avalanche/v1:default  # for v1 PCBs
    make avalanche/v2:default  # for v2 PCBs
    make avalanche/v3:default  # for v3 PCBs
```

File: keyboards/avalanche/readme.md [hunk 9]
```diff
@@ -4,17 +4,17 @@
...
 
 Flashing example for this keyboard:
 
-    make avalanche:default_v3:flash
+    make avalanche/v3:default:flash
```
Comment (noroadsleft) [inline keyboards/avalanche/readme.md:17]:
```suggestion
    make avalanche/v1:default:flash  # for v1 PCBs
    make avalanche/v2:default:flash  # for v2 PCBs
    make avalanche/v3:default:flash  # for v3 PCBs
```

File: keyboards/avalanche/v1/keymaps/default/keymap.c [hunk 10]
```diff
@@ -1,4 +1,4 @@
-// Copyright 2022 vitvlkv (@vlkv)
+// Copyright 2022 Viatly Volkov (@vlkv)
```
Comment (noroadsleft) [inline keyboards/avalanche/v1/keymaps/default/keymap.c:1]:
```suggestion
// Copyright 2022 Vitaly Volkov (@vlkv)
```

File: keyboards/avalanche/v1/v1.c [hunk 11]
```diff
@@ -1,4 +1,4 @@
-// Copyright 2022 vitvlkv (@vlkv)
+// Copyright 2022 Viatly Volkov (@vlkv)
```
Comment (noroadsleft) [inline keyboards/avalanche/v1/v1.c:1]:
```suggestion
// Copyright 2022 Vitaly Volkov (@vlkv)
```

File: keyboards/avalanche/v2/keymaps/default/keymap.c [hunk 12]
```diff
@@ -1,4 +1,4 @@
-// Copyright 2022 vitvlkv (@vlkv)
+// Copyright 2022 Viatly Volkov (@vlkv)
```
Comment (noroadsleft) [inline keyboards/avalanche/v2/keymaps/default/keymap.c:1]:
```suggestion
// Copyright 2022 Vitaly Volkov (@vlkv)
```

File: keyboards/avalanche/v2/v2.c [hunk 13]
```diff
@@ -1,4 +1,4 @@
-// Copyright 2022 vitvlkv (@vlkv)
+// Copyright 2022 Viatly Volkov (@vlkv)
```
Comment (noroadsleft) [inline keyboards/avalanche/v2/v2.c:1]:
```suggestion
// Copyright 2022 Vitaly Volkov (@vlkv)
```

File: keyboards/avalanche/v3/keymaps/default/keymap.c [hunk 14]
```diff
@@ -1,4 +1,4 @@
-// Copyright 2022 vitvlkv (@vlkv)
+// Copyright 2022 Viatly Volkov (@vlkv)
```
Comment (noroadsleft) [inline keyboards/avalanche/v3/keymaps/default/keymap.c:1]:
```suggestion
// Copyright 2022 Vitaly Volkov (@vlkv)
```

File: keyboards/avalanche/v3/v3.c [hunk 15]
```diff
@@ -1,4 +1,4 @@
-// Copyright 2022 vitvlkv (@vlkv)
+// Copyright 2022 Viatly Volkov (@vlkv)
```
Comment (noroadsleft) [inline keyboards/avalanche/v3/v3.c:1]:
```suggestion
// Copyright 2022 Vitaly Volkov (@vlkv)
```

File: keyboards/avalanche/v1/rules.mk [hunk 16]
```diff
@@ -0,0 +1,5 @@
+MCU = atmega32u4
+BOOTLOADER = caterina
+SPLIT_KEYBOARD = yes
+
+DEFAULT_FOLDER = avalanche/v1
```
Comment (noroadsleft) [inline keyboards/avalanche/v1/rules.mk:5]:
```suggestion
```

File: keyboards/avalanche/v2/rules.mk [hunk 17]
```diff
@@ -0,0 +1,7 @@
+MCU = atmega32u4
+BOOTLOADER = caterina
+RGBLIGHT_ENABLE = yes       # Enable WS2812 RGB underlight.
+SPLIT_KEYBOARD = yes
+ENCODER_ENABLE = yes
+
+DEFAULT_FOLDER = avalanche/v2
```
Comment (noroadsleft) [inline keyboards/avalanche/v2/rules.mk:7]:
```suggestion
```

File: keyboards/avalanche/v3/rules.mk [hunk 18]
```diff
@@ -0,0 +1,7 @@
+MCU = atmega32u4
+BOOTLOADER = caterina
+RGBLIGHT_ENABLE = no
+SPLIT_KEYBOARD = yes
+ENCODER_ENABLE = yes
+
+DEFAULT_FOLDER = avalanche/v3
```
Comment (noroadsleft) [inline keyboards/avalanche/v3/rules.mk:7]:
```suggestion
```

**Review by noroadsleft — CHANGES_REQUESTED** [review]
Fixes some typos, polishing the readme, and removing redundant code.

**Comment by winder** [comment]
Thanks for the pro tip, I was wondering why "add to batch" was grayed out.

File: keyboards/avalanche/v2/rules.mk [hunk 19]
```diff
@@ -0,0 +1,5 @@
+MCU = atmega32u4
+BOOTLOADER = caterina
+RGBLIGHT_ENABLE = yes       # Enable WS2812 RGB underlight.
```
Comment (fauxpark) [inline keyboards/avalanche/v2/rules.mk:3]:
```suggestion
RGBLIGHT_ENABLE = yes
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

**Comment by winder** [comment]
PR is rebased on master.

File: keyboards/avalanche/v1/keymaps/default/config.h [hunk 20]
```diff
@@ -0,0 +1,6 @@
+// Copyright 2022 Vitaly Volkov (@vlkv)
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+#pragma once
+
+#define USE_SERIAL
```
Comment (drashna) [inline keyboards/avalanche/v1/keymaps/default/config.h:6]: This is the default behavior and isn't technically needed.  Nor is the rest of the fine, in that case.

File: keyboards/avalanche/v2/keymaps/default/keymap.c [hunk 21]
```diff
@@ -0,0 +1,72 @@
...
+        }
+    } else if (index == 1) { /* Second encoder */
+        if (clockwise) {
+            tap_code(KC_PGUP);
+        } else {
+            tap_code(KC_PGDOWN);
+        }
+    }
+    return true;
+}
```
Comment (drashna) [inline keyboards/avalanche/v2/keymaps/default/keymap.c:72]:
```suggestion
```

File: keyboards/avalanche/avalanche.c [hunk 22]
```diff
@@ -0,0 +1,4 @@
+// Copyright 2022 Vitaly Volkov (@vlkv)
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+#include "avalanche.h"
```
Comment (drashna) [inline keyboards/avalanche/avalanche.c:4]: This allows for a default behavior to be defined, without the need for the user to specify the behavior.  This is very useful for the configurator which doesn't have a way to configure this behavior, currently.

```suggestion
#include "avalanche.h"

#ifdef ENCODER_ENABLE
bool encoder_update_kb(uint8_t index, bool clockwise) {
    if (!encoder_update_user(index, clockwise)) { return false; }
    if (index == 0) {
        if (clockwise) {
            tap_code_delay(KC_VOLU, 10);
        } else {
            tap_code_delay(KC_VOLD, 10);
        }
    } else if (index == 1) { /* Second encoder */
        if (clockwise) {
            tap_code(KC_PGUP);
        } else {
            tap_code(KC_PGDOWN);
        }
    }
    return true;
}
#endif

```

File: keyboards/avalanche/v1/rules.mk [hunk 23]
```diff
@@ -0,0 +1,3 @@
+MCU = atmega32u4
+BOOTLOADER = caterina
+SPLIT_KEYBOARD = yes
```
Comment (drashna) [inline keyboards/avalanche/v1/rules.mk:3]:
```suggestion
# MCU name
MCU = atmega32u4

# Bootloader selection
BOOTLOADER = caterina

# Build Options
#   change yes to no to disable
#
BOOTMAGIC_ENABLE = yes      # Enable Bootmagic Lite
MOUSEKEY_ENABLE = yes       # Mouse keys
EXTRAKEY_ENABLE = yes       # Audio control and System control
CONSOLE_ENABLE = no         # Console for debug
COMMAND_ENABLE = no         # Commands for debug and configuration
NKRO_ENABLE = yes           # Enable N-Key Rollover
BACKLIGHT_ENABLE = no       # Enable keyboard backlight functionality
RGBLIGHT_ENABLE = no        # Enable keyboard RGB underglow
AUDIO_ENABLE = no           # Audio output

SPLIT_KEYBOARD = yes
```

File: keyboards/avalanche/v2/keymaps/default/config.h [hunk 24]
```diff
@@ -0,0 +1,7 @@
+// Copyright 2022 Vitaly Volkov (@vlkv)
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+#pragma once
+
+#define USE_SERIAL
```
Comment (drashna) [inline keyboards/avalanche/v2/keymaps/default/config.h:6]: this file should be removed.

File: keyboards/avalanche/v2/rules.mk [hunk 25]
```diff
@@ -0,0 +1,5 @@
+MCU = atmega32u4
+BOOTLOADER = caterina
+RGBLIGHT_ENABLE = yes
+SPLIT_KEYBOARD = yes
+ENCODER_ENABLE = yes
```
Comment (drashna) [inline keyboards/avalanche/v2/rules.mk:5]:
```suggestion
# MCU name
MCU = atmega32u4

# Bootloader selection
BOOTLOADER = caterina

# Build Options
#   change yes to no to disable
#
BOOTMAGIC_ENABLE = yes      # Enable Bootmagic Lite
MOUSEKEY_ENABLE = yes       # Mouse keys
EXTRAKEY_ENABLE = yes       # Audio control and System control
CONSOLE_ENABLE = no         # Console for debug
COMMAND_ENABLE = no         # Commands for debug and configuration
NKRO_ENABLE = yes           # Enable N-Key Rollover
BACKLIGHT_ENABLE = no       # Enable keyboard backlight functionality
RGBLIGHT_ENABLE = yes       # Enable keyboard RGB underglow
AUDIO_ENABLE = no           # Audio output
ENCODER_ENABLE = yes

SPLIT_KEYBOARD = yes
```

File: keyboards/avalanche/v3/keymaps/default/config.h [hunk 26]
```diff
@@ -0,0 +1,7 @@
+// Copyright 2022 Vitaly Volkov (@vlkv)
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+#pragma once
+
+#define USE_SERIAL
```
Comment (drashna) [inline keyboards/avalanche/v3/keymaps/default/config.h:6]: This file should be removed.

File: keyboards/avalanche/v3/rules.mk [hunk 27]
```diff
@@ -0,0 +1,5 @@
+MCU = atmega32u4
+BOOTLOADER = caterina
+RGBLIGHT_ENABLE = no
+SPLIT_KEYBOARD = yes
+ENCODER_ENABLE = yes
```
Comment (drashna) [inline keyboards/avalanche/v3/rules.mk:5]:
```suggestion
# MCU name
MCU = atmega32u4

# Bootloader selection
BOOTLOADER = caterina

# Build Options
#   change yes to no to disable
#
BOOTMAGIC_ENABLE = yes      # Enable Bootmagic Lite
MOUSEKEY_ENABLE = yes       # Mouse keys
EXTRAKEY_ENABLE = yes       # Audio control and System control
CONSOLE_ENABLE = no         # Console for debug
COMMAND_ENABLE = no         # Commands for debug and configuration
NKRO_ENABLE = yes           # Enable N-Key Rollover
BACKLIGHT_ENABLE = no       # Enable keyboard backlight functionality
RGBLIGHT_ENABLE = no        # Enable keyboard RGB underglow
AUDIO_ENABLE = no           # Audio output
ENCODER_ENABLE = yes

SPLIT_KEYBOARD = yes
```

File: keyboards/avalanche/avalanche.c (hunk 22 above)
Comment (winder) [reply keyboards/avalanche/avalanche.c:4]: Thanks, I was wondering about a way to share code like this but was only able to find the userspace feature. Is there something similar for sharing oled screen behavior?

Comment (drashna) [reply keyboards/avalanche/avalanche.c:4]: > Is there something similar for sharing oled screen behavior?

Yup!  `oled_task_kb`, and return false on `!oled_task_user()`. Eg

```c
bool oled_task_kb(void) {
    if (!oled_task_user()) {
        return false;
    }
    // run keyboard code
}
```

**Review by drashna — APPROVED** [review]

**Comment by winder** [comment]
@fauxpark @noroadsleft could this get merged in? Thanks!

**Comment by noroadsleft** [comment]
Thanks!

**Pull request merged**